flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/QuickSort.java

```diff
@@ -45,24 +55,61 @@ protected static int getMaxDepth(int x) {
 
 	/**
 	 * Sort the given range of items using quick sort. {@inheritDoc} If the recursion depth falls below
-	 * {@link #getMaxDepth},
-	 * then switch to {@link HeapSort}.
+	 * {@link #getMaxDepth}, then switch to {@link HeapSort}.
 	 */
 	public void sort(final IndexedSortable s, int p, int r) {
-		sortInternal(s, p, r, getMaxDepth(r - p));
+		int recordsPerSegment = s.recordsPerSegment();
+		int recordSize = s.recordSize();
+
+		int maxOffset = recordSize * (recordsPerSegment - 1);
+
+		int size = s.size();
+		int sizeN = size / recordsPerSegment;
+		int sizeO = (size % recordsPerSegment) * recordSize;
+
+		sortInternal(s, recordsPerSegment, recordSize, maxOffset, 0, 0, 0, size, sizeN, sizeO, getMaxDepth(r - p));
 	}
 
 	public void sort(IndexedSortable s) {
 		sort(s, 0, s.size());
 	}
 
-	private static void sortInternal(final IndexedSortable s, int p, int r, int depth) {
+	/**
+	 * Sort the given range of items using quick sort. If the recursion depth falls below
+	 * {@link #getMaxDepth}, then switch to {@link HeapSort}.
+	 *
+	 * @param s paged sortable
+	 * @param recordsPerSegment number of records per memory segment
+	 * @param recordSize number of bytes per record
+	 * @param maxOffset offset of a last record in a memory segment
+	 * @param p index of first record in range
+	 * @param pN page number of first record in range
+	 * @param pO page offset of first record in range
+	 * @param r index of last-plus-one'th record in range
+	 * @param rN page number of last-plus-one'th record in range
+	 * @param rO page offset of last-plus-one'th record in range
+	 * @param depth recursion depth
+	 *
+	 * @see #sort(IndexedSortable, int, int)
+	 */
+	private static void sortInternal(final IndexedSortable s, int recordsPerSegment, int recordSize, int maxOffset,
+			int p, int pN, int pO, int r, int rN, int rO, int depth) {
 		while (true) {
 			if (r - p < 13) {
```

Comment:
Maybe you could add a comment here saying "switch to insertion sort", because it is now not that obvious from looking at the code.